src/librustc_parse/parser/ty.rs

```diff
@@ -276,8 +276,33 @@ impl<'a> Parser<'a> {
     }
 
     fn parse_borrowed_pointee(&mut self) -> PResult<'a, TyKind> {
-        let opt_lifetime = if self.check_lifetime() { Some(self.expect_lifetime()) } else { None };
+        let mut opt_lifetime =
+            if self.check_lifetime() { Some(self.expect_lifetime()) } else { None };
         let mutbl = self.parse_mutability();
+        if mutbl == Mutability::Mut && opt_lifetime.is_none() && self.token.is_lifetime() {
+            // A lifetime is invalid here: it would be part of a bare trait bound, which requires
+            // it to be followed by a plus, but we disallow plus in the pointee type.
+            // So we can handle this case as an error here, and suggest `'a mut`.
+            // If there *is* a plus next though, handling the error later provides better suggestions
+            // (like adding parentheses)
+            if !self.look_ahead(1, |t| t.is_like_plus()) {
+                let span = self.prev_token.span.to(self.token.span);
+                let lifetime = self.expect_lifetime();
+
+                let mut err = self.struct_span_err(span, "lifetime must precede `mut`");
+                if self.span_to_snippet(span).is_ok() {
+                    err.span_suggestion(
+                        span,
+                        "place the lifetime before `mut`",
+                        format!("{} mut", lifetime),
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+                err.emit();
+
+                opt_lifetime = Some(lifetime);
```

Comment:
```suggestion
                self.expect_lifetime();

                opt_lifetime = Some(lifetime);
```
Can you skip the lifetime token here? This should prevent the IMO unhelpful later errors.

Reply:
This is done above (L290, `let lifetime = self.expect_lifetime();`) in the case where we do emit an error here.

In the cases where we decide not to emit an error here, that's because I'm aiming for this suggestion:
```
LL | fn y<'a>(y: &mut 'a + Send) {
   |             ^^^^^^^^^^^^^^ help: try adding parentheses: `&mut ('a + Send)`
```

Which is the only thing I think an author could have meant in this case (why else the `+`, which is what determines whether we error now or not)?

Unfortunately we don't get that suggestion in expression contexts, and I agree the "at least one trait is required for an object type" is noise. (The bare trait object warning is correct, and would still occur if the author wrote `&mut ('a + Send)`). But these aren't the focus of the PR.

Reply:
You're right :+1: my bad, I looked at `y<'a>(y: &mut 'a + Send)` which still has a lot of errors, while `x<'a>(x: &mut 'a i32){}` is the one we care about here.